File: client/src/app/core/api/project-query.ts
import { HttpParams } from '@angular/common/http';

import { APP_CONFIG } from '@core/config/app-config';
import type { IsoDate } from '@core/models/project';

// Sortable columns. Mirrors the server's ProjectSort allow-list exactly.
export const PROJECT_SORT = {
  name: 'name',
  projectCode: 'projectCode',
  projectType: 'projectTypeName',
  status: 'status',
  startDate: 'startDate',
  endDate: 'endDate',
  activityCount: 'activityCount',
  durationDays: 'durationDays',
  lastModifiedOn: 'lastModifiedOn',
  lastModifiedBy: 'lastModifiedByName',
} as const;

export type ProjectSortKey = (typeof PROJECT_SORT)[keyof typeof PROJECT_SORT];
export type SortDirection = 'asc' | 'desc';

// A map extent as [minLon, minLat, maxLon, maxLat] in WGS84.
export type Bbox = readonly [number, number, number, number];

// Every input the Projects List can vary, in one object. Twin of the server's ProjectQuery.
export interface ProjectQuery {
  readonly page: number;
  readonly pageSize: number;
  readonly sort: ProjectSortKey;
  readonly dir: SortDirection;
  readonly search: string;
  readonly typeIds: readonly number[];
  // Status *names* (`InProgress`), which is what the server parses.
  readonly statuses: readonly string[];
  readonly dateFrom: IsoDate | null;
  readonly dateTo: IsoDate | null;
  // Set by the "only projects in current map extent" toggle.
  readonly bbox: Bbox | null;
  // Set by the Sketch spatial filter. WKT polygon in WGS84.
  readonly polygonWkt: string | null;
}

export const EMPTY_QUERY: ProjectQuery = {
  page: 1,
  pageSize: APP_CONFIG.list.pageSize,
  sort: PROJECT_SORT.lastModifiedOn,
  dir: 'desc',
  search: '',
  typeIds: [],
  statuses: [],
  dateFrom: null,
  dateTo: null,
  bbox: null,
  polygonWkt: null,
};

// True when anything beyond paging and sorting is narrowing the result set.
export function hasActiveFilters(query: ProjectQuery): boolean {
  return (
    query.search.trim().length > 0 ||
    query.typeIds.length > 0 ||
    query.statuses.length > 0 ||
    query.dateFrom !== null ||
    query.dateTo !== null ||
    query.bbox !== null ||
    query.polygonWkt !== null
  );
}

// The single place a query becomes parameters, so the export cannot drift from the grid.
export function toHttpParams(query: ProjectQuery, includePaging = true): HttpParams {
  let params = new HttpParams().set('sort', query.sort).set('dir', query.dir);

  if (includePaging) {
    params = params.set('page', query.page).set('pageSize', query.pageSize);
  }

  const search = query.search.trim();
  if (search) {
    params = params.set('search', search);
  }

  for (const id of query.typeIds) {
    params = params.append('typeIds', id);
  }

  for (const status of query.statuses) {
    params = params.append('statuses', status);
  }

  if (query.dateFrom) params = params.set('dateFrom', query.dateFrom);
  if (query.dateTo) params = params.set('dateTo', query.dateTo);

  if (query.bbox) {
    const [minLon, minLat, maxLon, maxLat] = query.bbox;
    params = params
      .set('minLon', minLon)
      .set('minLat', minLat)
      .set('maxLon', maxLon)
      .set('maxLat', maxLat);
  }

  if (query.polygonWkt) {
    params = params.set('polygonWkt', query.polygonWkt);
  }

  return params;
}

File: server/PMGIS.Infrastructure/Projects/ProjectQuery.cs
using PMGIS.Domain.Enums;

namespace PMGIS.Infrastructure.Projects;

// Every input the Projects List can vary, in one object.
public sealed record ProjectQuery
{
    public int Page { get; init; } = 1;
    public int PageSize { get; init; } = 25;

    // One of ProjectSort.
    public string Sort { get; init; } = ProjectSort.LastModifiedOn;
    public bool Descending { get; init; } = true;

    // Free-text, matched against Project Name and Project Code.
    public string? Search { get; init; }

    public IReadOnlyList<int> ProjectTypeIds { get; init; } = [];
    public IReadOnlyList<ProjectStatus> Statuses { get; init; } = [];

    // Inclusive range.
    public DateOnly? DateFrom { get; init; }
    public DateOnly? DateTo { get; init; }

    // Map extent for the "only projects in current map extent" toggle.
    public double? MinLongitude { get; init; }
    public double? MinLatitude { get; init; }
    public double? MaxLongitude { get; init; }
    public double? MaxLatitude { get; init; }

    // Polygon drawn with the Sketch widget, as WKT in WGS84.
    public string? PolygonWkt { get; init; }

    public bool HasExtent =>
        MinLongitude.HasValue && MinLatitude.HasValue &&
        MaxLongitude.HasValue && MaxLatitude.HasValue;
}

// Allow-list of sortable columns.
public static class ProjectSort
{
    public const string Name = "name";
    public const string ProjectCode = "projectCode";
    public const string ProjectType = "projectTypeName";
    public const string Status = "status";
    public const string StartDate = "startDate";
    public const string EndDate = "endDate";
    public const string ActivityCount = "activityCount";
    public const string DurationDays = "durationDays";
    public const string LastModifiedOn = "lastModifiedOn";
    public const string LastModifiedBy = "lastModifiedByName";

    public static readonly string[] All =
    [
        Name, ProjectCode, ProjectType, Status, StartDate, EndDate,
        ActivityCount, DurationDays, LastModifiedOn, LastModifiedBy,
    ];

    public static bool IsValid(string? sort) =>
        sort is not null && All.Contains(sort, StringComparer.OrdinalIgnoreCase);

    // Maps user input onto the exact constant, case-insensitively.
    public static string Normalise(string? sort) =>
        All.FirstOrDefault(s => string.Equals(s, sort, StringComparison.OrdinalIgnoreCase))
        ?? LastModifiedOn;
}
